File: ml/pricing/reloop_pricing/pricing/seasonality.py
"""Seasonality index — 1:1 mirror of apps/api/src/lib/seasonality.ts (spec 024,
phase 2). Keep the category buckets and monthly curves in lock-step with that
file, same convention as significance.py mirroring events.ts. Unlike geo.py
(which needs a cohort demand signal from the marketplace world), this needs no
external state at all — real Indian retail seasonality (Diwali, wedding
season, back-to-school) is stable enough to encode directly.
"""

from __future__ import annotations

from datetime import date
from typing import Dict, List

CATEGORY_ALIASES: Dict[str, str] = {
    "electronics": "electronics",
    "mobile": "electronics",
    "cell phones": "electronics",
    "fashion": "fashion",
    "apparel": "fashion",
    "clothing": "fashion",
    "women": "fashion",
    "men": "fashion",
    "home": "home",
    "furniture": "home",
    "home & garden": "home",
    "toys": "toys",
    "sports": "sports",
    "books": "books",
    "books_media": "books",
    "media": "books",
    "beauty": "beauty",
}

# Index 0 = January … 11 = December. Identical numbers to seasonality.ts.
SEASONAL_CURVE: Dict[str, List[float]] = {
    "electronics": [1.1, 1.05, 0.95, 0.9, 0.9, 1.0, 1.05, 1.1, 1.0, 1.2, 1.25, 1.1],
    "fashion": [1.15, 1.1, 0.95, 0.95, 1.0, 1.0, 0.9, 0.9, 0.95, 1.15, 1.2, 1.15],
    "home": [1.0, 1.0, 0.95, 0.95, 0.95, 1.0, 1.0, 1.0, 1.0, 1.15, 1.2, 1.1],
    "toys": [0.95, 0.9, 0.9, 0.95, 1.0, 1.05, 1.0, 0.95, 0.95, 1.1, 1.25, 1.2],
    "sports": [1.0, 1.0, 1.05, 1.05, 1.0, 0.95, 0.9, 0.9, 0.95, 1.0, 1.05, 1.05],
    "books": [1.0, 1.0, 1.05, 1.1, 1.05, 1.1, 1.0, 0.95, 1.0, 1.0, 1.0, 1.0],
    "beauty": [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.95, 0.95, 1.0, 1.15, 1.2, 1.1],
    "other": [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.05, 1.1, 1.05],
}

SCORE_MIN = 0.7
SCORE_MAX = 1.3


def clamp(value: float, lo: float, hi: float) -> float:
    return min(hi, max(lo, value))


def get_seasonality_index(category: str, at: date | None = None) -> float:
    bucket = CATEGORY_ALIASES.get(str(category).strip().lower(), "other")
    at = at or date.today()
    raw = SEASONAL_CURVE[bucket][at.month - 1]
    return clamp(raw, SCORE_MIN, SCORE_MAX)

File: apps/api/src/lib/seasonality.ts
// Seasonality index — a deterministic, calendar-driven category × month curve
// (spec 024, phase 2). Before this, PricingStateVector.seasonalityIndex was a
// named field with no seasonal logic at all: always the flat 0.5 placeholder
// in reprice-engine.ts's fillState(), and a random draw in the Python
// marketplace simulator. Unlike geoDemandIndex (spec 024, phase A), this
// signal needs no live event aggregation or database — real Indian retail
// seasonality (Diwali, wedding season, back-to-school) is well-known and
// stable enough to encode directly, so this is a pure function, always
// available even when Mongo is unconfigured.

type CategoryBucket = 'electronics' | 'fashion' | 'home' | 'toys' | 'sports' | 'books' | 'beauty' | 'other';

// PricingStateVector.category is a free-form string sourced from either the
// Sell-flow ItemCategory union (electronics/home/fashion/sports/toys/books/
// other) or the eBay/Mercari training catalogue's category strings (women/
// men/vintage & collectibles/etc, see ml/pricing's CATEGORY_BIAS) — this alias
// table normalizes both onto the same seasonal buckets.
const CATEGORY_ALIASES: Record<string, CategoryBucket> = {
  electronics: 'electronics',
  mobile: 'electronics',
  'cell phones': 'electronics',
  fashion: 'fashion',
  apparel: 'fashion',
  clothing: 'fashion',
  women: 'fashion',
  men: 'fashion',
  home: 'home',
  furniture: 'home',
  'home & garden': 'home',
  toys: 'toys',
  sports: 'sports',
  books: 'books',
  books_media: 'books',
  media: 'books',
  beauty: 'beauty',
};

function resolveCategoryBucket(category: string): CategoryBucket {
  return CATEGORY_ALIASES[category.trim().toLowerCase()] ?? 'other';
}

// One multiplier per calendar month (index 0 = January … 11 = December),
// centered on 1.0 = neutral. Same clamp range as demand_index (computeDemandIndex.ts)
// for consistency — this feature and geoDemandIndex share one scale.
const SEASONAL_CURVE: Record<CategoryBucket, readonly number[]> = {
  // Republic Day sale (Jan), Independence Day sale (Aug), Diwali (Oct/Nov).
  electronics: [1.1, 1.05, 0.95, 0.9, 0.9, 1.0, 1.05, 1.1, 1.0, 1.2, 1.25, 1.1],
  // Wedding season (Jan/Feb, Nov/Dec), Diwali (Oct/Nov).
  fashion: [1.15, 1.1, 0.95, 0.95, 1.0, 1.0, 0.9, 0.9, 0.95, 1.15, 1.2, 1.15],
  // Diwali home-refresh + wedding-season gifting.
  home: [1.0, 1.0, 0.95, 0.95, 0.95, 1.0, 1.0, 1.0, 1.0, 1.15, 1.2, 1.1],
  // Diwali + New Year gifting.
  toys: [0.95, 0.9, 0.9, 0.95, 1.0, 1.05, 1.0, 0.95, 0.95, 1.1, 1.25, 1.2],
  sports: [1.0, 1.0, 1.05, 1.05, 1.0, 0.95, 0.9, 0.9, 0.95, 1.0, 1.05, 1.05],
  // Academic-year start (Apr/Jun) bump.
  books: [1.0, 1.0, 1.05, 1.1, 1.05, 1.1, 1.0, 0.95, 1.0, 1.0, 1.0, 1.0],
  beauty: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.95, 0.95, 1.0, 1.15, 1.2, 1.1],
  other: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.05, 1.1, 1.05],
};

const SCORE_MIN = 0.7;
const SCORE_MAX = 1.3;

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

/** Deterministic seasonal demand index for a category at a given date (default
 *  now). No Mongo dependency — always real, unlike the geo-demand feature. */
export function getSeasonalityIndex(category: string, at: Date = new Date()): number {
  const bucket = resolveCategoryBucket(category);
  const raw = SEASONAL_CURVE[bucket][at.getMonth()]!;
  return clamp(raw, SCORE_MIN, SCORE_MAX);
}
